pkg/scheduler/node_iterator.go

```diff
@@ -0,0 +1,106 @@
+package scheduler
+
+import (
+	"math/rand"
+)
+
+// NodeIterator iterates over a list of nodes based on the defined policy
+type NodeIterator interface {
+	// returns true if there are more values to iterate over
+	HasNext() (ok bool)
+	// returns the next node from the iterator
+	Next() (node *SchedulingNode)
+	// reset the iterator to a clean state
+	Reset()
+}
+
+// All iterators extend the base iterator
+type baseIterator struct {
+	NodeIterator
+	countIdx int
+	size     int
+	nodes    []*SchedulingNode
+}
+
+// Reset the iterator to start from the beginning
+func (bi *baseIterator) Reset() {
+    bi.countIdx = 0
+}
+
+// HasNext returns true if there is a next element in the array.
+// Returns false if there are no more elements or list is empty.
+func (bi *baseIterator) HasNext() bool {
+	if (bi.countIdx + 1) > bi.size {
+		return false
+	}
+	return true
+}
+
+// Next returns the next element and advances to next element in array.
+// Returns nil and reset the starting point at the end of iteration.
+func (bi *baseIterator) Next() *SchedulingNode {
+	if (bi.countIdx + 1) > bi.size {
+		return nil
+	}
+
+	value := bi.nodes[bi.countIdx]
+	bi.countIdx++
+	return value
+}
+
+// Default iterator, wraps the base iterator.
+// Iterates over the list from the start, position zero, to end.
+type DefaultNodeIterator struct {
+	baseIterator
+}
+
+// Create a new default iterator
+func NewDefaultNodeIterator(schedulerNodes []*SchedulingNode) *DefaultNodeIterator {
+	it := &DefaultNodeIterator{}
+	it.nodes = schedulerNodes
+	it.size = len(schedulerNodes)
+	return it
+}
+
+// Random iterator, wraps the base iterator
+// Iterates over the list from a random starting position in the list.
+// The iterator automatically wraps at the end of the list.
+type RandomNodeIterator struct {
```

Comment:
The name is a bit confusing. RandomNodeIterator to me sounds like it picks up a random node every time. But the actual, it is a round-robin starting from a random point. Can we call this `RoundRobinNodeIterator`?

Reply:
ah yes that is much better